src/utils/mod.rs

```diff
@@ -274,6 +308,117 @@ impl TpmsRsaParmsBuilder {
+    /// Asymmetric scheme to be used for key operations
+    pub scheme: AsymSchemeUnion,
+    /// Curve to be used with the key
+    pub curve: EllipticCurve,
+    /// Flag indicating whether the key shall be used for signing
+    pub for_signing: bool,
+    /// Flag indicating whether the key shall be used for decryption
+    pub for_decryption: bool,
+    /// Flag indicating whether the key is restricted
+    pub restricted: bool,
+}
+
+impl TpmsEccParmsBuilder {
+    /// Create parameters for a restricted decryption key (i.e. a storage key)
+    pub fn new_restricted_decryption_key(symmetric: Cipher, curve: EllipticCurve) -> Self {
+        TpmsEccParmsBuilder {
+            symmetric: Some(symmetric),
+            scheme: AsymSchemeUnion::AnySig(None),
+            curve,
+            for_signing: false,
+            for_decryption: true,
+            restricted: true,
+        }
+    }
+
+    /// Create parameters for an unrestricted signing key
+    pub fn new_unrestricted_signing_key(scheme: AsymSchemeUnion, curve: EllipticCurve) -> Self {
+        TpmsEccParmsBuilder {
+            symmetric: None,
+            scheme,
+            curve,
+            for_signing: true,
+            for_decryption: false,
+            restricted: false,
+        }
+    }
+
+    /// Build an object given the previously provded parameters.
+    ///
+    /// The only mandatory parameter is the asymmetric scheme.
+    ///
+    /// # Errors
+    /// * if no asymmetric scheme is set, `ParamsMissing` wrapper error is returned.
+    /// * if the `for_signing`, `for_decryption` and `restricted` parameters are
+    /// inconsistent with the rest of the parameters, `InconsistentParams` wrapper
+    /// error is returned
+    pub fn build(self) -> Result<TPMS_ECC_PARMS> {
+        if self.restricted && self.for_decryption {
+            if self.symmetric.is_none() {
+                return Err(Error::local_error(WrapperErrorKind::ParamsMissing));
+            }
+        } else if self.symmetric.is_some() {
+            return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+        }
+        if self.for_decryption && self.for_signing {
+            return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+        }
+
+        let scheme = self.scheme.get_ecc_scheme_struct();
+        if self.for_signing
+            && scheme.scheme != TPM2_ALG_ECDSA
+            && scheme.scheme != TPM2_ALG_ECDAA
+            && scheme.scheme != TPM2_ALG_SM2
+            && scheme.scheme != TPM2_ALG_ECSCHNORR
+        {
+            return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+        }
+
+        if self.for_decryption
+            && scheme.scheme != TPM2_ALG_SM2
+            && scheme.scheme != TPM2_ALG_ECDH
+            && scheme.scheme != TPM2_ALG_ECMQV
+            && scheme.scheme != TPM2_ALG_NULL
+        {
+            return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+        }
```

Comment:
I ask this mostly because I am not an expert but it seems like here we are checking that the appropriate algorithm is select for each key type. Does that mean that different algorithms are used for signing and verification?

Reply:
No, `for_signing` actually means `for_signing_and_verification` and decryption is the same with encryption. It essentially specifies what the private key is used for. 

This split here is because some scheme are for signing and others are for encryption. `SM2` can do both